src/utilities/hoist.ts

```diff
@@ -0,0 +1,47 @@
+const REACT_LIFECYCLE_EXCLUSIONS = [
+  'setState',
+  'forceUpdate',
+  'render',
+  'componentWillMount',
+  'componentDidMount',
+  'componentWillReceiveProps',
+  'shouldComponentUpdate',
+  'componentWillUpdate',
+  'componentDidUpdate',
+  'componentWillUnmount'
+];
+
+/**
+ * Allows you to hoist methods, except those in an exclusion set from a source object into a destination object.
+ * @returns {string[]} An array of names of methods that were hoisted.
+ */
+export function hoistMethods(destination, source, exclusions = REACT_LIFECYCLE_EXCLUSIONS): string[] {
+  let hoisted: string[] = [];
+  for (let methodName in source) {
+    if (
+      typeof source[methodName] === 'function' &&
+      destination[methodName] === undefined &&
+      exclusions.indexOf(methodName) === -1
+    ) {
+      hoisted.push(methodName);
+      /* tslint:disable:no-function-expression */
+      destination[methodName] = function () { source[methodName].apply(source, arguments); };
+      /* tslint:enable */
+    }
+  }
+
+  return hoisted;
+}
+
+/**
+ * Provides a method for convenience to unhoist hoisted methods.
+ * @param {Object} source The source object upon which methods were hoisted.
+ * @param {string[]} methodNames An array of method names to unhoist.
+ */
+export function unhoistMethods(source: Object, methodNames: string[]): void {
+  for (let methodName of methodNames) {
+    if (source.hasOwnProperty(methodName) && typeof source[methodName] === 'function') {
+      source[methodName] = null; // or should this be undefined?
```

Comment:
Hi David,


Would it be OK for Cliff to check in now so I can consume the API in sp-client. This change for our WEXFR bug which needs to get check in by end of today. Cliff will continue refactoring the code to move to base component. But this will unblock me from making changes on my end.


Thanks,

Ping
_
ø¤º°`°º¤ø,¸¸,ø¤º°`°`° ø¤º°`°º¤ø,¸¸,ø¤º°`°`°ø¤º°`°º¤ø,¸¸,
Hello - Hola - Ni Hao - Bonjour - Konichiwa - World Readiness
¤º°`°º¤ø,¸¸,ø¤º°`°`° ø¤º°`°º¤ø,¸¸,ø¤º°`°`°ø¤º°`°º¤ø,¸¸,
________________________________
From: David Zearing <notifications@github.com>
Sent: Tuesday, August 30, 2016 11:22:28 AM
To: OfficeDev/office-ui-fabric-react
Cc: Ping Jiang; Assign
Subject: Re: [OfficeDev/office-ui-fabric-react] Add hoisting to all decorators (#218)


In src/utilities/hoist.ts<https://github.com/OfficeDev/office-ui-fabric-react/pull/218#discussion_r76851068>:

> +      /* tslint:enable */
> +    }
> +  }
> +
> +  return hoisted;
> +}
> +
> +/**
> + * Provides a method for convenience to unhoist hoisted methods.
> + * @param {Object} source The source object upon which methods were hoisted.
> + * @param {string[]} methodNames An array of method names to unhoist.
> + */
> +export function unhoistMethods(source: Object, methodNames: string[]): void {
> +  for (let methodName of methodNames) {
> +    if (source.hasOwnProperty(methodName) && typeof source[methodName] === 'function') {
> +      source[methodName] = null; // or should this be undefined?


delete source[methodName]

-
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub<https://github.com/OfficeDev/office-ui-fabric-react/pull/218/files/758094eaffad0a21789fb55f67c2a78420ad8cb8#r76851068>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AD0L_UWSOypD6w_VeehExfGsT0U8vjfCks5qlHTkgaJpZM4Jwui2>.